Framework/no implicit import paths

dmsnell opened the pull request:
My IDE has trouble with our implicit import statements that rely on node resolution for `client`

In this PR I have created a codemod to transform those implicit imports into `client/`-prefixed imports.

We only transform if:
 - the path doesn't start with `.`
 - the path has a `/` in it
 - the path resolves to a real folder or `.js` or `.jsx` file in `client/`

**Testing**
Smoke test. We need some verification that the appropriate files are updated and no incorrect modules are transformed.

Run with `bin/codemods/run.js no-implicit-import-paths client/` to transform and watch for the modules which _weren't_ transformed, such as `lodash/noop` or `react-dom/server`

samouri commented on bin/codemods/src/no-implicit-import-paths.js:
```diff
@@ -0,0 +1,65 @@
+		.filter( dec => {
+			const path = dec.value.source.value;
+
+			const isDot = path.startsWith( '.' );
```
`isDot` --> `isRelative`

dmsnell replied:
done

samouri commented on bin/codemods/src/no-implicit-import-paths.js:
```diff
@@ -0,0 +1,65 @@
+			return Object.assign(
+				j.importDeclaration(
+					node.value.specifiers,
+					j.literal( node.value.source.value.replace( /^/, 'client/' ) )
```
TIL you can prepend by replacing the `/^/`! Thats pretty cool.

I wonder if it would read simpler as:
`j.literal( 'client/' + node.source.value )`

dmsnell replied:
done - I have no idea what I was thinking

sirreal commented:
If I've understood the description correctly, these imports are under `client`, in use, and wouldn't be caught because they don't include `/`:

```
config
dispatcher
extensions
notices
sections
sections-preload
state
```

Want to add explicit matches?

simison commented:
Nice!

You could drop the `/`-requirement by checking if import is external. See [sort-imports](https://github.com/Automattic/wp-calypso/blob/fbf7c23a68dc74760f8fc6a3aa0377c476b4f786/bin/codemods/src/sort-imports.js#L42) for an example.

ockham commented:
Does this enable us to drop `modules` from https://github.com/Automattic/wp-calypso/blob/6078cfb61d2d51adc6f3d73c0ceda994b11e2ee8/webpack.config.js#L127 and https://github.com/Automattic/wp-calypso/blob/6078cfb61d2d51adc6f3d73c0ceda994b11e2ee8/webpack.config.node.js#L113-L119? ([ref](https://webpack.js.org/configuration/resolve/#resolve-modules)).

dmsnell commented:
> Does this enable us to drop modules

Probably but I'd prefer to do that in a separate PR just because right now this is (theoretically) a _safe_ patch because it continues to work as it has without any behavioral changes. (Actually it doesn't build but that's another problem).

After the merge I'd like to circle back to our configs and find where we can eliminate the work-arounds.

> You could drop the /-requirement by checking if import is external.

Not sure I'd trust that; I'd rather know we have a valid importable module from `client` - on that note, I believe that I _can_ remove the `/` but can probably just leave out any further checks.

> If I've understood the description correctly, these imports are under client, in use, and wouldn't be caught because they don't include /:

see above: we probably don't need any explicit matching; just check if they exist

nylen commented:
> Other tools also have problems with this system too.

Any examples?

I thought any tool that can understand the `client/...` syntax is already using non-standard functionality because by default, imports are relative to the current file or `node_modules`, not the to project root.  I wonder if these tools (and your IDE) could be told to treat `client/` as the "project root" instead?

> Does this enable us to drop modules from ...

Wouldn't we need to instead change `path.join( __dirname, 'client' )` to just `__dirname` because, again, we're still using non-standard syntax?  Wouldn't this also need to be done in this PR to make the build work?

ockham commented:
> I wonder if these tools (and your IDE) could be told to treat `client/` as the "project root" instead?

But that's going to cause issues in other directories (`server/` specifically), isn't it?

> because by default, imports are relative to the current file or `node_modules`, not the to project root.

TBH I wasn't aware.

dmsnell commented:
> imports are relative to the current file or node_modules, not the to project root

this is true, which is why `client/` works, because `client` is at the top-level with `package.json` where `node` is being run. `state` or `components` however are implicit in how they are actually located inside a subdirectory only knowable from parsing `webpack.config.js`.

> I wonder if these tools (and your IDE) could be told to treat client/ as the "project root" instead?

Well I swear I couldn't do this before when I tried but just now while making sure I didn't make a fool of myself by disagreeing with you publicly @nylen I found that I could add the `client` directory as a `Source` directory in PhpStorm and that makes the connections work. :wink:

> But that's going to cause issues in other directories (server/ specifically), isn't it?

Good point @ockham; and even the `config` module I had to exclude on account of the fact that the paths inside our modules are actually making certain assumptions. There aren't too many of them, as seen here in the code for this codemod, but still they are there and my preference would be to remove that implicit stuff.

---

@nylen I'm not sure I fully understand the force behind your comment. Are you opposed to making this change or specifically addressing some of the points I raised?

nylen commented:
> I'm not sure I fully understand the force behind your comment. Are you opposed to making this change or specifically addressing some of the points I raised?

Sorry @dmsnell, I meant to investigate more and respond to this earlier.

My concerns with this PR are as follows:

> this is true, which is why `client/` works, because `client` is at the top-level with `package.json` where `node` is being run.

According to my understanding of import rules, and quick testing, this doesn't seem to be correct.  Simply changing imports from `thing` to `client/thing` does not work without further changes:

```
ERROR in ./client/state/ui/post-type-list/reducer.js
Module not found: Error: Can't resolve 'client/state/action-types' in '.../wp-calypso/client/state/ui/post-type-list'
```

Vanilla Node says the same thing:

```sh
wp-calypso $ node
> require( 'client/state' )
Error: Cannot find module 'client/state'
```

This is also a big (in LOC terms) change that touches every part of a mature codebase and requires all developers to start using a new convention immediately, so I think there is a good amount of risk and not much gain here.

dmsnell commented:
Mentioning #12303 here since it may be related that our implicit paths are making it harder for tools like `eslint` to verify our imports are valid.

granted, as soon as we eliminate commonJS this won't be a problem because webpack will complain

samouri commented:
which is very soon :) https://github.com/Automattic/wp-calypso/pull/16057
